File: hasAllCodes/fastestSolution.java
// link: https://leetcode.com/problems/check-if-a-string-contains-all-binary-codes-of-size-k/

class Solution {
    public boolean hasAllCodes(String s, int k) {
        if (s.length() < (1 << k) + k - 1) {
            return false;
        }
        int n = 0;
        char[] chars = s.toCharArray();
        for (int i = 0; i < k - 1; i++) {
            n = (n << 1) + (chars[i] - '0');
        }
        int size = 1 << k;
        int mask = size - 1;
        boolean[] exists = new boolean[size];
        for (int i = k - 1; i < chars.length; i++) {
            n = ((n << 1) + chars[i] - '0') & mask;
            if (!exists[n]) {
                exists[n] = true;
                if (--size == 0) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: removeAllAdjancentDuplicatesInStringII/fastestSolution.cpp
// link: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string-ii/
class Solution {
public:
	string removeDuplicates(string s, int k) {
		bool replaced = true;

		while (replaced) {
			replaced = false;
			int cnt = 1;

			for (int i = 1; i < s.size(); ++i) {
				if (s[i] == s[i-1]) {
					if (++cnt == k)	{
						s = s.substr(0, i - k + 1) + s.substr(i + 1);
						replaced = true;
					}
				}
				else cnt = 1;
			}
		}

		return s;
	}
}